rfc/merkle-structure.md

```diff
@@ -0,0 +1,505 @@
+
+## Goals
+
+Now that we have outlined problems, we can enumerate what is desired to improve upon and why:
+
+1. Data structure SHOULD be identified by the same content identifier regardless of how parts its comprised of are stored. It SHOULD NOT matter if some parts are local and other parts external. This will get us flexibility to make choices about which parts to inline vs externalize without affecting data identifier. It will give us flexibility to make different choices based on storage constraints and possibly different choices during data transfer.
+
+1. Data structure SHOULD be identified by the same content identifier regardless of the data encoding. This will give us flexibility to choose encoding based on storage constraint and ability to negotiate content encoding during data transfer.
+
+## Approach
+
+At the high level we propose to identify data structures by the root of the [reference tree], which is a binary merkle tree deterministically derived from the data structure itself.
+
+### Data Types
+
+[IPLD data model] offers same data types as pretty much every mainstream programming language. We use same set of data types except for the [IPLD Link], which is obsolete since every piece of data can be canonically identified.
+
+### Reference Tree
+
+Reference tree is a binary [merkle tree] deterministically derived from the data structure. It is derived according to the following algorithm:
+
+1. [Abstract data format], list of two or more nodes, is derived from the data structure.
+1. First node, an **operator**, denotes relation the rest of nodes, **operands** form. It is kept around for the final step.
+1. Operands are [merkle fold]ed into an **operand** (sub)tree.
+1. **Operator** node and **operand** (sub)tree are [merkle fold]ed into a tree that represents **reference tree** of the data.
+
+### Merkle Fold
+
+Merkle fold is a process of deriving [BAO] inspired binary [merkle tree] from the list of leaf nodes. Parent nodes in derived tree have exactly two children and the content of each parent node is the hash of its left child concatenated with the hash of its right child. When there's an odd number of nodes in a given level of the tree, the rightmost node is raised to the level above.
+
+```mermaid
+graph TB
+parent(root) --> A(leaf)
+parent --> B(leaf)
+
+parent3(root) --> parent2(node)
+parent2 --> a(leaf)
+parent2 --> b(leaf)
+parent3 --> c(leaf)
+
+parent4(root) --> parent5(node)
+parent4 --> parent6(node)
+parent5 --> ch51(leaf)
+parent5 --> ch52(leaf)
+parent6 --> ch61(leaf)
+parent6 --> ch62(leaf)
+```
+
+Merkle fold of a single leaf is a leaf itself
+
+```mermaid
+graph TB
+only(leaf)
+``````
+
+Merkle fold of zero leaves is an empty hash (hash of zero bytes)
```

Comment:
```suggestion
Merkle fold of zero leaves is a leaf that is byte array with 0 bytes.
```

I think this might be a better choice. Consequence is that empty list becomes double hash of the list operator. Also null abstract data format can be defined as null tag and empty tree.